core/src/main/java/com/linecorp/armeria/common/Flags.java

```diff
@@ -88,279 +80,228 @@
 import io.netty.util.ReferenceCountUtil;
 
 /**
- * The system properties that affect Armeria's runtime behavior.
+ * The system properties that affect Armeria's runtime behavior. The priority or each flag is determined
+ * by {@link FlagsProvider#priority()}. On each flag, if value is fail to validated. The next candidate will be
+ * used.
  */
 public final class Flags {
 
     private static final Logger logger = LoggerFactory.getLogger(Flags.class);
 
-    private static final Splitter CSV_SPLITTER = Splitter.on(',').trimResults().omitEmptyStrings();
-
     private static final String PREFIX = "com.linecorp.armeria.";
 
-    private static final int NUM_CPU_CORES = Runtime.getRuntime().availableProcessors();
+    private static final List<FlagsProvider> FLAGS_PROVIDER =
+            Lists.newArrayList(ServiceLoader.load(FlagsProvider.class, Flags.class.getClassLoader()))
```

Comment:
We prefer to use Guava's immutable collections.
```suggestion
            ImmutableList.copyOf(ServiceLoader.load(FlagsProvider.class, Flags.class.getClassLoader()))
```

Reply:
Shouldn't we always add `DefaultFlagsProvider` to the last and `SystemPropertyFlagsProvider` to the first?
```
SystemPropertyFlagsProvider -> (sorted) custom A FlagsProvider -> custom B FlagsProvider -> ... -> DefaultFlagsProvider
```

Reply:
Previous I forgot to `git add` META-INF/services 😂. Anyway, I changed flags providers as you suggested.